### PR TITLE
Allow specifying expected IL verification baseline

### PR DESCRIPTION
**Pull request by tmat**
Currently a test can only specify whether IL verification should pass or fail. 
This change allows it to specify what exactly causes the IL verification to fail, so that the test can define list of expected IL verification violations and guard against unexpected ones.

### CODE REVIEW

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (jcouv) [inline src/Compilers/Test/Core/CommonTestBase.cs:42]: Do we plan to use this type for value equality? If not, let's not use a record.

File: src/Compilers/Test/Core/Platform/Desktop/DesktopRuntimeEnvironment.cs
Comment (jcouv) [inline src/Compilers/Test/Core/Platform/Desktop/DesktopRuntimeEnvironment.cs:343]: I don't think we should do this substitution on the expected message

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (jcouv) [inline src/Compilers/Test/Core/CommonTestBase.cs:42]: Please update a test to show how `ILVerifyMessage` and `PEVerifyMessage` are used in practice (with non-default values) and to exercise the new comparison logic.

File: src/Compilers/Test/Core/Platform/Desktop/DesktopRuntimeEnvironment.cs
Comment (tmat) [reply src/Compilers/Test/Core/Platform/Desktop/DesktopRuntimeEnvironment.cs:341]: It's deterministic, but the tokens are not interesting since the message already contains the qualified member name.
The tokens change if you change the code in any way, so you can't have multiple tests reusing the same baseline PE verify messages.

File: src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs:10333]: Consider also updating a test for ILVerifyMessage.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2) modulo CI failure (missing ctor)

File: src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs
Comment (tmat) [reply src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs:10333]: IL verify does not report an error here.

Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs:10333]: Any test where IL verify complains will do.

Comment (tmat) [reply src/Compilers/CSharp/Test/Emit/CodeGen/UnsafeTests.cs:10333]: I see what you mean

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (333fred) [inline src/Compilers/Test/Core/CommonTestBase.cs:44]: `required`?

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (333fred) [inline src/Compilers/Test/Core/CompilationVerifier.cs:341]: I would prefer:
```cs
$"""
IL Verify failed unexpectedly:
{actualMessage}
"""
```

File: src/Compilers/Test/Core/CommonTestBase.cs
Comment (tmat) [reply src/Compilers/Test/Core/CommonTestBase.cs:44]: Not really required. `default(Verification)` is intentionally equivalent to `Pass`.

**Review by 333fred — APPROVED** [review]

File: src/Compilers/Test/Core/Assert/ConditionalFactAttribute.cs
Comment (jcouv) [inline src/Compilers/Test/Core/Assert/ConditionalFactAttribute.cs:233]: We don't use target-typed new in the compiler codebase, so that we can do textual search for `new Type`

Comment (jcouv) [inline src/Compilers/Test/Core/Assert/ConditionalFactAttribute.cs:233]:
```suggestion
        public static readonly IsEnglishLocal Instance = new IsEnglishLocal();
```

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 6)

File: src/Compilers/Test/Core/CompilationVerifier.cs
Comment (tmat) [reply src/Compilers/Test/Core/CompilationVerifier.cs]: Verbatim string formats really badly in the ternary expression, but I'm gonna use interpolation instead.

**Comment by 333fred** [comment]
@tmat please wait for 2 approvals before merging.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 9)

**Pull request merged**
